text/0121-data-types.md

```diff
@@ -0,0 +1,435 @@
+
+User document field order can be lost if the engine flattens identical field values are not co-located in a document payload.
+
+The following JSON document:
+
+```json
+{
+    "a.b": "T-shirt",
+    "price": 2.0,
+    "a": {
+        "b": "Nice T-shirt"
+    }
+}
+```
+
+Is internally represented
+
+```json
+{
+    "a.b": ["T-shirt", "Nice T-shirt"],
+    "price": 2.0
+}
+```
+
+The second representation of `a.b` in its nested form is merged with the first representation of `a.b`.
+
+Users can't and should not rely on a given document field order when `searchableAttributes` is `[*]`.
+
+#### 3.2.4.7. Dot-notation Expression
+
+Permits to express the nested object property.
+
+##### 3.2.4.7.2. Example
+
+Given this document structure
+
+```json
+{
+    "id": 0,
+    "person": {
+        "firstname": "John",
+        "lastname": "Doe",
+        "address": {
+            "country": "US",
+            "city": "New York"
+        }
+    }
+}
+```
+
+A precise field can be expressed using the dot-notation
+
+```json
+{
+    "attributesToHighlight": ["person.firstname"]
+}
+```
+
+#### 3.2.4.8. All Object Properties Expression
+
+It is possible to express the definition of all properties of an object.
+
+e.g. In this case `person` is an object containing properties. `attributesToRetrieve: ["person"]`
+
+This notation is accepted on all parameters or settings allowing to specify attributes. This is due to the fact that several documents may not share the same schema. See [3.2.4.8.3. Edge Case](#32483-edge-case) section.
+
+##### 3.2.4.8.2. Example
+
+Given this document structure
+
+```json
+{
+    "id": 0,
+    "person": {
+        "firstname": "John",
+        "lastname": "Doe",
+        "address": {
+            "country": "US",
+            "city": "New York"
+        }
+    }
+}
+```
+
+All properties of a document nested object can be expressed this way
+
+```json
+{
+    "attributesToHighlight": ["person"]
+}
+```
+
+It's equivalent to
+
+```json
+{
+    "attributesToHighlight": [
+        "person.firstname",
+        "person.lastname",
+        "person.address.country",
+        "person.address.city"
+    ]
+}
+```
+
+##### 3.2.4.8.3. Edge Case
+
+One document might have a non-nested field `person` while another has a `person` object containing properties.
+
+The chosen behavior is not to force the user's hand and Meilisearch do not throw any errors.
+
+E.g. If a user specifies `"filter": "person = 'Guillaume'"` at search time, the document that would have a nested object person would not be brought up by the filter.
+
+## 4. Future Possibilities
+
+- Change the default behavior of `searchableAttributes` so that it is predictable. We may remove the priority based on a field position in a document.
+- Support the wildcard notation with the dot-notation. e.g. `person.*`, `person.address.*` or `person.l*`
+- Support the array notation. e.g. `person.addresses[1]`
```

Comment:
A big ➕ 1️⃣ for changing the default behavior of `searchableAttributes`. Right now it seems to be somewhere between defined and undefined behavior (how Meilisearch represents and orders document attributes internally _is_ presumably defined, but we are not sharing that definition w/ the user 😏).

Reply:
Yes, it also bothers me not to be able to explain it clearly or to have to say "undefined behavior". I think we can work on it the day we allow to define the same weight between two fields.